weave/flow/eval_imperative.py

```diff
@@ -0,0 +1,325 @@
+            BeforeValidator(_cast_to_cls(Scorer)),
+            Field(
+                description="A metadata-only scorer used for comparisons."
+                "Alternatively, you can pass a dict of attributes or just a string"
+                "representing the ID of your scorer."
+            ),
+        ],
+        score: ScoreType,
+    ) -> None:
+        # this is safe; pydantic casting is done in validator above
+        scorer = cast(Scorer, scorer)
+
+        @weave.op(name=scorer.name)
+        def score_method(self: Scorer, *, output: Any, **inputs: Any) -> ScoreType:
+            # TODO: can't use score here because it will cause version mismatch
+            # return score
+            return cast(ScoreType, current_score.get())
+
+        scorer.__dict__["score"] = MethodType(score_method, scorer)
+
+        # attach the score feedback to the predict call
+        with call_context.set_call_stack(
+            [self.evaluate_call, self.predict_and_score_call]
+        ):
+            with _set_current_score(score):
+                await self.predict_and_score_call.apply_scorer(scorer)
+
+
+class ImperativeEvaluationLogger(BaseModel):
+    """This class provides an imperative interface for logging evaluations.
+
+    An evaluation is started automatically when the first prediction is logged
+    using the `log_prediction` method, and finished when the `log_summary` method
+    is called.
+
+    Each time you log a prediction, you will get back an `ImperativePredictionLogger`
+    object.  You can use this object to log scores and metadata for that specific
+    prediction (see that class for more details).
+
+    Example:
+        ```python
+        ev = ImperativeEvaluationLogger()
+        pred = ev.log_prediction(inputs, output)
+        pred.log_score(scorer_name, score)
+        ev.log_summary(summary)
+        ```
+    """
+
+    model: Annotated[
+        Model | dict | str,
+        BeforeValidator(_cast_to_cls(Model)),
+        Field(
+            default_factory=Model,
+            description="A metadata-only Model used for comparisons."
+            "Alternatively, you can pass a dict of attributes or just a string"
+            "representing the ID of your model.",
+        ),
+    ]
+
+    _eval_started: bool = PrivateAttr(False)
+    _logged_summary: bool = PrivateAttr(False)
+    _evaluate_call: Call | None = PrivateAttr(None)
+    _pseudo_evaluation: Evaluation = PrivateAttr(
+        default_factory=lambda: Evaluation(
+            dataset=weave.Dataset(rows=weave.Table([{"": ""}])),
+            scorers=[],
+        )
+    )
+
+    @model_validator(mode="after")
+    def _validate_model(self) -> ImperativeEvaluationLogger:
+        if isinstance(self.model, str):
+            # Convert string to a model instance
+            # Create a dynamic model class with the string as the name
+            model_name = self.model
+            DynamicModel = type(model_name, (Model,), {})
```

Comment:
don't we need to handle the dict case here?

Reply:
thanks for catching -- this is extraneous and now removed.  All casting is done in the annotation above.